<!--
Copyright [2009-2014] EMBL-European Bioinformatics Institute
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
     http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

{% extends "portal/base.html" %}
{% load staticfiles %}
{% load humanize %}
{% load portal_extras %}

{% block meta_tags %}
    {{ block.super }}
    <meta name="description" content="Unique RNA sequence {{ rna.upi }}, {{ rna.length }} nucleotides long"/>
{% endblock %}

{% block title %}
{{ rna.upi }}
{% endblock %}

{% block content %}

<h1>
  Sequence {{ rna.upi }}

  {% if context.single_species %}
    <small><em>{{ context.single_species }}</em></small>
  {% else %}
    <small>{{ rna.count_distinct_organisms|intcomma }}</strong> species</small>
  {% endif %}

  <small class="pull-right hidden-xs">
    <button type="button" class="btn btn-info tour help" title="Take an interactive tour to explore RNAcentral features" id="guided-tour">
      <i class="fa fa-info-circle"></i> Interactive tour
    </button>
  </small>
</h1>

{% if context.taxid and not context.taxid_filtering %}
  <div class="alert alert-danger">
    You requested annotations from taxid:{{context.taxid}}, but none were found. Please check the taxid in the URL.
  </div>
{% elif context.taxid_filtering %}
  <div class="well well-sm">
    <i class="fa fa-filter"></i> Showing annotations from <strong><em>{{ context.single_species}}</em></strong>.

    {% if rna.count_distinct_organisms > 1 %}
      <a href="" class="show-species-tab">Switch species</a> or
      view data from {{ rna.count_distinct_organisms|add:"-1" }} <a href="{% url 'portal.views.rna_view' rna.upi %}">more species</a>.
    {% endif %}
  </div>
{% elif rna.count_distinct_organisms > 1 %}
  <div class="well well-sm">
    <i class="fa fa-info-circle"></i> This unique sequence was observed in multiple species.
    <a href="" class="show-species-tab">Filter annotations</a> by species.
  </div>
{% endif %}

<div class="row">

  <div class="col-md-12">

    <div class="tabbable">

      <ul class="nav nav-tabs" id="tabs">
        <li class="active">
          <a href="" data-target="#overview" data-toggle="tab">Overview</a>
        </li>
        <li>
          <a href="" data-target="#species" data-toggle="tab">Taxonomy</a>
        </li>
        <li>
          <a href="" data-target="#related-seqs" data-toggle="tab">Related human sequences</a>
        </li>
        <li class="disabled hidden-xs" data-toggle="tab">
          <a data-target="#2d" class="help" title="Coming soon">2D</a>
        </li>
        <li class="disabled hidden-xs" data-toggle="tab">
          <a data-target="#3d" class="help" title="Coming soon">3D</a>
        </li>
        <li class="dropdown pull-right hidden-xs">
          <a class="dropdown-toggle" data-toggle="dropdown" href="">
                      <i class="fa fa-download"></i> Download
                      <b class="caret"></b>
                  </a>
          <ul class="dropdown-menu" id="download-formats">
            <li><a href="/api/v1/rna/{{ rna.upi }}.json" download={{ rna.upi }}.json>json</a></li>
            <li><a href="/api/v1/rna/{{ rna.upi }}.yaml" download={{ rna.upi }}.yaml>yaml</a></li>
            <li><a href="/api/v1/rna/{{ rna.upi }}.fasta" download={{ rna.upi }}.fasta>fasta</a></li>
          </ul>
        </li>
      </ul>

      <div class="tab-content">

        <div class="tab-pane fade in active" id="overview">

          <h2>
            Overview
          </h2>

          <div class="panel panel-default">
            <div class="panel-body">
              <ul class="list-inline">
                <li class="lead">{{ context.description }}</li>
              </ul>
              <ul class="list-inline">
                <li><strong>{{ rna.length|intcomma }}</strong> nucleotides</li>
                <li>
                  {% with dbs=context.distinct_databases %}
                    <strong>{{ dbs|length|intcomma }}</strong> database{{ dbs|length|pluralize }}
                    <small>
                      {% for db in dbs %}
                        {% if forloop.first %}({% endif %}{{db}}{% if not forloop.last %},{% else %}){% endif %}
                      {% endfor %}
                    </small>
                  {% endwith %}
                </li>
                <li>
                  Found in <strong>{{ rna.count_distinct_organisms|intcomma }}</strong> <a href="" class="show-species-tab">species</a>
                </li>
                <li><small>imported <time datetime="{{ rna.first_seen|date:'Y-m-d' }}">{{ rna.first_seen|date:'d M Y' }}</time></small></li>
                <li><small>last updated <time datetime="{{ rna.last_seen|date:'Y-m-d' }}">{{ rna.last_seen|date:'d M Y' }}</time></small></li>
              </ul>
            </div>
          </div> <!-- .panel .panel-default -->

          <div>
            <h2>
              Annotations
              <small id="xrefs-datatables-counter">{{ rna.count_xrefs|intcomma }} total</small>

              <div class="btn-group btn-group-xs margin-left-5px btn-group-xref-sort" style="display:none">
                <button type="button" class="btn btn-default dropdown-toggle" data-toggle="dropdown">
                  Sort by<span class="caret"></span>
                </button>
                <ul class="dropdown-menu" role="menu">
                  <li><a id="sort-by-first-seen">first seen</a></li>
                  <li><a id="sort-by-last-seen">last seen</a></li>
                </ul>
              </div>

              <span id="xrefs-datatables-filter"></span>
            </h2>

            {% if rna.count_xrefs > context.xref_page_size %}
              <div class="alert alert-warning" role="alert">
                Showing {{ context.xref_page_size|intcomma }} entries out of {{ rna.count_xrefs|intcomma }}.
                Navigate between batches:
                <ul class="pagination pagination-sm">
                  {% for xref_page in context.xref_pages %}
                    <li {% if forloop.first %}class="active"{% endif %}>
                      <a href="" data-xref-page="{{xref_page}}" class="xref-pagination">{{ xref_page }}</a>
                    </li>
                  {% endfor %}
                </ul>
              </div>
            {% endif %}

            <div class="table-responsive force-scrollbars">

              <div class="panel panel-default" id="xrefs-loading">
                <div class="panel-body" id="xrefs-loading-msg">
                  <i class="fa fa-spinner fa-spin fa-2x"></i>
                  Loading database cross-references...
                </div>
              </div>

              <div id="xrefs-table-div"></div>

            </div> <!-- .table-responsive -->

            <div class="xref-datatables-controls"></div>
          </div>

          {# genoverse #}
          <div>
            <div class="wrap genoverse-wrap">
              <h2>
                Genome browser
                <small id="genoverse-coordinates"></small>
              </h2>
              <div id="genoverse-description"></div>
              <div id="genoverse"></div>
            </div>
          </div>


          <h2>
            Sequence
            <small>
              <object classid="clsid:d27cdb6e-ae6d-11cf-96b8-444553540000"
                      width="110"
                      height="14"
                      id="clippy" >
              <param name="movie" value="{% static "flash/clippy.swf" %}"/>
              <param name="allowScriptAccess" value="always" />
              <param name="quality" value="high" />
              <param name="scale" value="noscale" />
              <param NAME="FlashVars" value="text={{ rna.get_sequence }}">
              <param name="bgcolor" value="#FFFFFF">
              <embed src="{% static "flash/clippy.swf" %}"
                     width="110"
                     height="14"
                     name="clippy"
                     quality="high"
                     allowScriptAccess="always"
                     type="application/x-shockwave-flash"
                     pluginspage="http://www.macromedia.com/go/getflashplayer"
                     FlashVars="text={{ rna.get_sequence }}"
                     bgcolor="#FFFFFF"
              />
              </object>
            </small>
          </h2>

          <p>
            {{ rna.length|intcomma }} nucleotides
            <small>
              ({{ context.counts.A|intcomma }} A; {{ context.counts.C|intcomma }} C;
              {{ context.counts.G|intcomma }} G; {{ context.counts.U|intcomma }} U;
              {{ context.counts.N|intcomma|default:"no " }} N)
            </small>
          </p>
          <pre class="pre-scrollable">
{{ rna.get_sequence }}</pre>


        </div> <!-- .tab-pane .active -->

        <div class="tab-pane fade" id="species">

          <h2>Taxonomic Tree</h2>

          <p>
            This sequence has been observed in <strong>{{ rna.count_distinct_organisms }}</strong> organisms.
            Get <strong>species-specific urls</strong> by clicking on <span class="text-primary"><strong>species names</strong></span>.
            <br>
            <small class="text-muted">
              <i class="fa fa-info-circle"></i>
              <strong>Scroll around</strong> to explore the entire tree.
              <strong>Click</strong> tree nodes to collapse or expand them.
              <strong>Hover</strong> over taxon names to display additional information.
            </small>
          </p>

          <div class="panel panel-default d3-species">
            <div class="tab-pane panel-body d3-species force-scrollbars d3-species-tree-tab">
              <div id="d3-species-tree">
                <i class="fa fa-spinner fa-spin fa-2x"></i>
                Loading taxonomic tree...
              </div>
            </div>

          </div>

        </div>

        <div class="tab-pane" id="related-seqs">

          <button class="btn btn-default" ng-click="meta_search('({% for related in context.related_mcl_20 %}{{related.upi}}{% if not forloop.last %} OR {% endif %}{% endfor %})')">Browse related sequences</button>

          <h3>Overlapping accessions</h3>
          {% for overlap in rna.get_overlapping_entries %}
            <a href="/rna/{{ overlap }}">{{overlap}}</a>{% if not forloop.last %}, {% endif %}
          {% empty %}
            None found
          {% endfor %}

          <h3>Overlapping genomic locations</h3>

          <p>
            {% for overlap in rna.get_genomic_overlapping_entries %}
              <a href="/rna/{{ overlap.upi }}">{{overlap.upi}}</a>, {{overlap.precomputed.description}}{% if not forloop.last %}<br>{% endif %}
            {% empty %}
              None found
            {% endfor %}
          </p>

          <h3>MCL 6.0</h3>

          <table class="table table-hover table-condensed table-bordered" id="related-seqs-table2">
            <thead>
              <th>#</th>
              <th>RNAcentral id</th>
              <th>Length</th>
              <th>Description</th>
              <th>Human Xrefs</th>
              <th>Species</th>
              <th>Genome mapping</th>
              <th>Ns</th>
              <th>Databases</th>
              <th>Identity</th>
              <th>Hit length</th>
            </thead>

            {% for related in context.related_mcl_60 %}
              <tr>
                <td>{{ forloop.counter }}</td>
                <td>
                  {% if  related.upi == rna.upi %}
                    &rarr; <a href="/rna/{{ related.upi }}#homo">{{ related.upi }}</a></td>
                  {% else %}
                    <a href="/rna/{{ related.upi }}#homo">{{ related.upi }}</a></td>
                  {% endif %}
                <td>{{ related.length }}</td>
                <td>{{ related.precomputed.description }}</td>
                <td>{{ related.precomputed.count_human_xrefs }}</td>
                <td>{{ related.precomputed.count_distinct_organisms }}</td>
                <td>{{ related.precomputed.has_human_genomic_coordinates|yesno:"yes,no" }}</td>
                <td>{{ related.precomputed.N_symbols }}</td>
                <td>
                  {% with dbs=related.get_distinct_database_names %}
                    <small>
                      {% for db in dbs %}
                        {{db}}{% if not forloop.last %},{% endif %}
                      {% endfor %}
                    </small>
                  {% endwith %}
                </td>
                <td>{{ context.blast_identity|key_value:related.upi }}</td>
                <td>{{ context.blast_length|key_value:related.upi }}</td>
              </tr>
            {% endfor %}

          </table>

          <h3>MCL 2.0</h3>

          <table class="table table-hover table-condensed table-bordered" id="related-seqs-table-mcl-40">
            <thead>
              <th>#</th>
              <th>RNAcentral id</th>
              <th>Length</th>
              <th>Description</th>
              <th>Human Xrefs</th>
              <th>Species</th>
              <th>Genome mapping</th>
              <th>Ns</th>
              <th>Databases</th>
              <th>Identity</th>
              <th>Hit length</th>
            </thead>

            {% for related in context.related_mcl_20 %}
              <tr>
                <td>{{ forloop.counter }}</td>
                <td>
                  {% if  related.upi == rna.upi %}
                    &rarr; <a href="/rna/{{ related.upi }}#homo">{{ related.upi }}</a></td>
                  {% else %}
                    <a href="/rna/{{ related.upi }}#homo">{{ related.upi }}</a></td>
                  {% endif %}
                <td>{{ related.length }}</td>
                <td>{{ related.precomputed.description }}</td>
                <td>{{ related.precomputed.count_human_xrefs }}</td>
                <td>{{ related.precomputed.count_distinct_organisms }}</td>
                <td>{{ related.precomputed.has_human_genomic_coordinates|yesno:"yes,no" }}</td>
                <td>{{ related.precomputed.N_symbols }}</td>
                <td>
                  {% with dbs=related.get_distinct_database_names %}
                    <small>
                      {% for db in dbs %}
                        {{db}}{% if not forloop.last %},{% endif %}
                      {% endfor %}
                    </small>
                  {% endwith %}
                </td>
                <td>{{ context.blast_identity|key_value:related.upi }}</td>
                <td>{{ context.blast_length|key_value:related.upi }}</td>
              </tr>
            {% endfor %}

          </table>

          <h3>CD-HIT</h3>

          <p>Only the entries not clustered together by MCL</p>

          <table class="table table-hover table-condensed table-bordered" id="related-seqs-table">
            <thead>
              <th>#</th>
              <th>RNAcentral id</th>
              <th>Length</th>
              <th>Description</th>
              <th>Human Xrefs</th>
              <th>Species</th>
              <th>Genome mapping</th>
              <th>Ns</th>
              <th>Databases</th>
              <th>Identity</th>
            </thead>

            {% for related in rna.get_cdhit_related_entries %}
              <tr>
                <td>{{ forloop.counter }}</td>
                <td>
                  {% if  related.upi == rna.upi %}
                    &rarr; <a href="/rna/{{ related.upi }}#homo">{{ related.upi }}</a></td>
                  {% else %}
                    <a href="/rna/{{ related.upi }}#homo">{{ related.upi }}</a></td>
                  {% endif %}
                <td>{{ related.length }}</td>
                <td>{{ related.precomputed.description }}</td>
                <td>{{ related.precomputed.count_human_xrefs }}</td>
                <td>{{ related.precomputed.count_distinct_organisms }}</td>
                <td>{{ related.precomputed.has_human_genomic_coordinates|yesno:"yes,no" }}</td>
                <td>{{ related.precomputed.N_symbols }}</td>
                <td>
                  {% with dbs=related.get_distinct_database_names %}
                    <small>
                      {% for db in dbs %}
                        {{db}}{% if not forloop.last %},{% endif %}
                      {% endfor %}
                    </small>
                  {% endwith %}
                </td>
                <td>{{ context.blast_identity|key_value:related.upi }}</td>
              </tr>
            {% endfor %}

          </table>

          <h3>Blast</h3>

          <table class="table table-hover table-condensed table-bordered" id="related-seqs-table3">
            <thead>
              <th>#</th>
              <th>RNAcentral id</th>
              <th>Identity</th>
              <th>Length</th>
            </thead>

            {% for related in rna.get_blast_related_entries %}
              <tr>
                <td>{{ forloop.counter }}</td>
                <td>
                {% if related.query == related.target %}
                  &rarr; {{ related.target }}
                {% elif rna.upi == related.query %}
                  <a href="/rna/{{ related.target }}">{{ related.target }}</a>
                {% elif rna.upi == related.target %}
                  <a href="/rna/{{ related.query }}">{{ related.query }}</a>
                {% endif %}
                </td>
                <td>{{ related.identity }}</td>
                <td>{{ related.length }}</td>
              </tr>
            {% endfor %}

          </table>


        </div>

        <div class="tab-pane" id="2d">
        </div>

        <div class="tab-pane" id="3d">
        </div>

      </div> <!--  tab-content -->
    </div> <!--  tabbable -->
  </div> <!--  col-md-12 -->
</div> <!--  row -->

{# Handlebars #}
{% verbatim %}
<script id="handlebars-literature-reference-tmpl" type="text/x-handlebars-template">
  {{#with refs}}
    {{#each .}}
      <blockquote>
       <strong>
         {{title}}
       </strong>
       <br>
       {{authors}}
       <small>
        {{publication}}
        {{#if pubmed_id}}
          <div class="abstract-preview">
            <button class="btn btn-xs btn-default abstract-btn abstract-control" data-pubmed-id="{{pubmed_id}}">Show abstract</button>
            <a href="http://www.ncbi.nlm.nih.gov/pubmed/{{pubmed_id}}" target="_blank" class="abstract-control">PubMed</a>
            <div class="abstract-text"></div>
          </div>
        {{/if}}
       </small>
      </blockquote>
    {{/each}}
  {{/with}}
</script>

<script id="handlebars-loading-error-tmpl" type="text/x-handlebars-template">
  <div class='alert alert-danger'>
    <i class="fa fa-exclamation-triangle"></i>
    Sorry, there was a problem loading the data. Please try again and contact us if the problem persists.
  </div>
</script>

<script id="handlebars-loading-info-tmpl" type="text/x-handlebars-template">
  <div class='alert alert-info'>
    This may take a couple of minutes for sequences with a large number of cross-references.
  </div>
</script>

{% endverbatim %}


{% endblock content %}


{% block extra_js_uncompressed %}
  {{ block.super }}
    <script type="text/javascript">
    $(function() {
      {% if context.taxid_filtering %}
        rna_sequence_view = new rnaSequenceView('{{rna.upi}}', {{ context.taxid }});
      {% else %}
        rna_sequence_view = new rnaSequenceView('{{rna.upi}}');
      {% endif %}
      rna_sequence_view.initialize();

      // hopscotch guided tour
      hopscotch_tour = new guidedTour;
      hopscotch_tour.initialize();

      $('#related-seqs-table').dataTable({"aLengthMenu": [[5, 10, 20, 50, -1], [5, 10, 20, 50, "All"]]});
      $('#related-seqs-table2').dataTable({"aLengthMenu": [[5, 10, 20, 50, -1], [5, 10, 20, 50, "All"]]});
      $('#related-seqs-table3').dataTable({"aLengthMenu": [[5, 10, 20, 50, -1], [5, 10, 20, 50, "All"]]});
      $('#related-seqs-table-mcl-40').dataTable({"aLengthMenu": [[5, 10, 20, 50, -1], [5, 10, 20, 50, "All"]]});
    });
  </script>
{% endblock %}
